lib/glucose-get-last.js

```diff
@@ -1,329 +1,91 @@
-    return obj.glucose || obj.sgv;
-    }).map(function prepGlucose (obj) {
-        //Support the NS sgv field to avoid having to convert in a custom way
-        obj.glucose = obj.glucose || obj.sgv;
-        if ( obj.glucose !== null ) {
-            return obj;
-        }
-    });
+      var now = data[0];
+      var now_date = getDateFromEntry(now);
+      var change;
+      var last_deltas = [];
+      var short_deltas = [];
+      var long_deltas = [];
+      var last_cal = 0;
 
-    var now = data[0];
-    var now_date = getDateFromEntry(now);
-    var change;
-    var last_deltas = [];
-    var short_deltas = [];
-    var long_deltas = [];
-    var last_cal = 0;
+      //console.error(now.glucose);
+      for (var i=1; i < data.length; i++) {
+          // if we come across a cal record, don't process any older SGVs
+          if (typeof data[i] !== 'undefined' && data[i].type === "cal") {
+              last_cal = i;
+              break;
+          }
+          // only use data from the same device as the most recent BG data point
+          if (typeof data[i] !== 'undefined' && data[i].glucose > 38 && data[i].device === now.device) {
+              var then = data[i];
+              var then_date = getDateFromEntry(then);
+              var avgdelta = 0;
+              var minutesago;
+              if (typeof then_date !== 'undefined' && typeof now_date !== 'undefined') {
+                  minutesago = Math.round( (now_date - then_date) / (1000 * 60) );
+                  // multiply by 5 to get the same units as delta, i.e. mg/dL/5m
+                  change = now.glucose - then.glucose;
+                  avgdelta = change/minutesago * 5;
+              } else { console.error("Error: date field not found: cannot calculate avgdelta"); }
+              //if (i < 5) {
+                  //console.error(then.glucose, minutesago, avgdelta);
+              //}
+              // use the average of all data points in the last 2.5m for all further "now" calculations
+              if (-2 < minutesago && minutesago < 2.5) {
+                  now.glucose = ( now.glucose + then.glucose ) / 2;
+                  now_date = ( now_date + then_date ) / 2;
+                  //console.error(then.glucose, now.glucose);
+              // short_deltas are calculated from everything ~5-15 minutes ago
+              } else if (2.5 < minutesago && minutesago < 17.5) {
+                  //console.error(minutesago, avgdelta);
+                  short_deltas.push(avgdelta);
+                  // last_deltas are calculated from everything ~5 minutes ago
+                  if (2.5 < minutesago && minutesago < 7.5) {
+                      last_deltas.push(avgdelta);
+                  }
+                  //console.error(then.glucose, minutesago, avgdelta, last_deltas, short_deltas);
+              // long_deltas are calculated from everything ~20-40 minutes ago
+              } else if (17.5 < minutesago && minutesago < 42.5) {
```

Comment:
if `minutesago` is exactly 17.5 is this an issue as well? Same thing with the suggestion to include a `=` on one of these comparisons.

Reply:
I agree, at exactly 2.5 and 17.5min the delta assignment would be skipped